File: find-all-duplicates-in-an-array/find-all-duplicates-in-an-array.cpp
class Solution {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        
//        vector<int>v;
        
//        int correct_idx;
//        int i=0;

//         while (i<arr.size()) {
            
//             if(arr[i]!=i+1){
//             correct_idx=arr[i]-1;


//             if(arr[i]!=arr[correct_idx])
//             {
//                 int temp=arr[i];
//                 arr[i]=arr[correct_idx];
//                 arr[correct_idx]=temp;
//             }
//              else
//                     v.push_back(i);
//             }
//             else
//             i++;
//         }
        
//         return v;
        
         vector<int>p;
        sort(nums.begin(),nums.end());
        for(int i=1;i<nums.size();i++)
        {
            if(nums[i]==nums[i-1])
            {
                p.push_back(nums[i]);
            }
        }
        return p;
    }
};

File: find-all-duplicates-in-an-array/find-all-duplicates-in-an-array.java
class Solution {
    public List<Integer> findDuplicates(int[] arr) {
        
        int n=arr.length;
        List<Integer>l=new ArrayList<>();
        
        int correct_idx;
        int i=0;

        while (i<arr.length) {
            correct_idx=arr[i]-1;

            if(arr[i]!=arr[correct_idx])
            {
                int temp=arr[i];
                arr[i]=arr[correct_idx];
                arr[correct_idx]=temp;
            }
            else
            i++;
        }
        
        for(int j=0;j<n;j++)
        {
            if(arr[j]!=j+1)
                l.add(arr[j]);
        }
        return l;
    }
}